File: Neetcode/TrapRainWater.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size() - 1;
        int maxL = height[0];
        int maxR = height[n];
        int water = 0;
        int i=0, j=n;
        while (i < j){
            if (maxL <= maxR){
                i += 1;
                maxL = max(maxL, height[i]);
                water += maxL - height[i];
            }
            else{
                j -= 1;
                maxR = max(maxR, height[j]);
                water += maxR - height[j];
            }
        }
        return water;
    }
};

File: Neetcode/TrapRainWater.py
from collections import deque
class Solution:
    def trap(self, height: List[int]) -> int:
        leftMax = []
        rightMax = []
        for h in height:
            if not leftMax:
                leftMax.append(h)
            else:
                if leftMax[-1] <= h:
                    leftMax.append(h)
                else:
                    leftMax.append(leftMax[-1])
        for h in height[::-1]:
            if not rightMax:
                rightMax.append(h)
            else:
                if rightMax[-1] <= h:
                    rightMax.append(h)
                else:
                    rightMax.append(rightMax[-1])
        water = 0
        n = len(height) - 1 
        for i, h in enumerate(height):
            water += min(leftMax[i], rightMax[n-i]) - h
        return water
